source/extensions/path/pattern_template_lib/pattern_template.cc

```diff
@@ -0,0 +1,171 @@
+#include "source/extensions/path/pattern_template_lib/pattern_template.h"
+
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "source/common/http/path_utility.h"
+#include "source/extensions/path/pattern_template_lib/pattern_template_internal.h"
+#include "source/extensions/path/pattern_template_lib/proto/pattern_template_rewrite_segments.pb.h"
+
+#include "absl/status/statusor.h"
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "re2/re2.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace PatternTemplate {
+
+using PatternTemplateInternal::ParsedUrlPattern;
+
+#ifndef SWIG
+// Silence warnings about missing initializers for members of LazyRE2.
+#pragma GCC diagnostic ignored "-Wmissing-field-initializers"
+#endif
+
+inline re2::StringPiece toStringPiece(absl::string_view text) { return {text.data(), text.size()}; }
+
+absl::StatusOr<std::string> convertURLPatternSyntaxToRegex(absl::string_view url_pattern) {
+
+  absl::StatusOr<ParsedUrlPattern> status =
+      PatternTemplateInternal::parseURLPatternSyntax(url_pattern);
+  if (!status.ok()) {
+    return status.status();
+  }
+  struct ParsedUrlPattern pattern = *std::move(status);
+  return PatternTemplateInternal::toRegexPattern(pattern);
+}
+
+absl::StatusOr<std::vector<RewritePatternSegment>>
+parseRewritePatternHelper(absl::string_view pattern) {
+  std::vector<RewritePatternSegment> result;
+
+  // Don't allow contiguous '/' patterns.
+  static const LazyRE2 invalid_regex = {"^.*//.*$"};
+  if (RE2::FullMatch(toStringPiece(pattern), *invalid_regex)) {
+    return absl::InvalidArgumentError("Invalid rewrite literal pattern");
+  }
+
+  // The pattern should start with a '/' and thus the first segment should
+  // always be a literal.
+  if (pattern.empty() || pattern[0] != '/') {
+    return absl::InvalidArgumentError("Invalid rewrite variable placement");
+  }
+  while (!pattern.empty()) {
+    std::vector<absl::string_view> segments1 = absl::StrSplit(pattern, absl::MaxSplits('{', 1));
+    if (!segments1[0].empty()) {
+      if (!PatternTemplateInternal::isValidRewriteLiteral(segments1[0])) {
+        return absl::InvalidArgumentError("Invalid rewrite literal pattern");
+      }
+      result.emplace_back(segments1[0], RewriteStringKind::KLiteral);
+    }
+
+    if (segments1.size() < 2) {
```

Comment:
What happens if the pattern is "/foo{"? It looks a bit like this will split into "/foo" and "", so we'll simply return the "/foo" literal but ignore the trailing "{". Is that right?

Reply:
Let me write a test for it and get back to you.

Reply:
Your above logic is correct. We could have a rewrite is the middle of url segment like this: /abc-def-{var1}/a/{var1}"